### ISSUE TITLE
Minification fails if there is whitespace before an opening tag whose end tag is (legally) omitted

### ISSUE THREAD

**Issue by chris-morgan** [wilsonzlin/minify-html#22]
If there’s any whitespace before the first opening tag in a document for an element where you should be allowed to omit the end tag, it fails:

```console
$ # Space after the doctype
$ minify-html-cli <<< '<!doctype html> <html>'
Failed at character 23:
Expected closing tag.
--------
1|<!doctype html><html>>
2|
>|R

$ minify-html-cli <<< '<!doctype html> <p>'
Failed at character 20:
Expected closing tag.
--------
1|<!doctype html><p>>
2|
>|R

$ # Space before the doctype
$ minify-html-cli <<< ' <!doctype html><html>'
Failed at character 23:
Expected closing tag.
--------
1|<!doctype html><html>>
2|
>|R

$ minify-html-cli <<< ' <!doctype html><p>'
Failed at character 20:
Expected closing tag.
--------
1|<!doctype html><p>>
2|
>|R

$ # No doctype
$ minify-html-cli <<< ' <html>'
Failed at character 8:
Expected closing tag.
--------
1|<html>>
2|
>|R

$ minify-html-cli <<< ' <p>'
Failed at character 5:
Expected closing tag.
--------
1|<p>>
2|
>|R
```

But you’re fine if there’s no space before the first opening tag:

```console
$ minify-html-cli <<< '<!doctype html><html>'
<!doctype html><html>
$ minify-html-cli <<< '<!doctype html><p>'
<!doctype html><p>
$ minify-html-cli <<< '<html>'
<html>
$ minify-html-cli <<< '<p>'
<p>
```

(P.S. Why are the `<html>`, `<head>` and `<body>` start tags always retained? They should be omitted where possible for better compression. But that’s probably a separate issue.)

You’re also fine if you provide the end tag (and in the output it trims the leading whitespace and the end tag if acceptable):

```console
$ minify-html-cli <<< '<!doctype html> <html></html>'
<!doctype html><html>
$ minify-html-cli <<< '<!doctype html> <p></p>'
<!doctype html><p>
$ minify-html-cli <<< ' <html></html>'
<html>
$ minify-html-cli <<< ' <p></p>'
<p>
```

I hit this immediately in trying out minify-html, because my document had a line feed between the doctype and the html start tag, and I omitted the html end tag:

```html
<!doctype html>
<html lang=en-AU>
<meta charset=utf-8>
<title>…</title>
And some body, but no end tag for the html element.
```

**Comment by chris-morgan** [comment]
Just broadened the title of the issue a little after realising it’s broader than just the first tag in the document:

```console
$ minify-html-cli <<< '<!doctype html><html><div><p>Foo</div></html>'
<!doctype html><html><div><p>Foo</div>
$ minify-html-cli <<< '<!doctype html><html><div> <p>Foo</div></html>'
Failed at character 38:
Closing tag name does not match opening tag (expected "p", got "div").
--------
1|<!doctype html><html><div><p>Fooo</div></html>
>|                                      R
```

Yeah, I like to omit various optional end tags in my hand-written HTML.

**Comment by wilsonzlin** [comment]
Thanks for raising and appreciate the detailed description and examples. I've just found the cause of the bug, and it should be fixed now in the latest version `0.4.3`; let me know if it works for you. It was using an overwritten value for checking the tag name, which was why it wasn't working when the HTML was shifted down (from trimming the leading whitespace).

Regarding the `head` and `body` opening tag omission: when I looked into it a while ago, the official rules were quite complex due to the fact that nodes (like `<title>`, `<script>`, `<meta>`, and text nodes) must be carefully handled to not "leak" from `head` to `body` (and vice versa); since minify-html aims processes in-place and in one pass, it was tricky to implement the "technically correct" behaviour. However I'll revisit this again soon, as I think it should be safe for most realistic usages anyway, and could be put behind a flag for extra safety.
